feat(allocator): `FixedSizeAllocator` store flag recording if owned by both Rust and JS

overlookmotel opened the pull request:
Add a field `is_double_owned` to `FixedSizeAllocatorMetadata`. This field tracks ownership of the allocator/buffer. It's `true` if both Rust and JS currently hold a reference to the memory, or `false` if only one does.

`Drop` impl for `FixedSizeAllocator` checks the value of this field and only frees the memory if JS doesn't also hold a reference.

At present, `is_double_owned` is always `false`, because we're not transferring buffers over to JS yet. But it'll come into play in later PR. At that point the finalizer which runs when JS garbage collector collects the buffer will also update `is_double_owned`, and free the memory if Rust already dropped the `FixedSizeAllocator`, so JS is the only remaining owner.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12381?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12382** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12382?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12381** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12381?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12381?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12310** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12310?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12380** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12380?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12379** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12379?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12330** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12330?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#12348](https://github.com/oxc-project/oxc/pull/12348) <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12348?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#12374** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12374?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12381).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented:
### Merge activity

* **Jul 18, 8:52 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
feat(allocator): `FixedSizeAllocator` store flag recording if owned by both Rust and JS (#12381)

Add a field `is_double_owned` to `FixedSizeAllocatorMetadata`. This field tracks ownership of the allocator/buffer. It's `true` if both Rust and JS currently hold a reference to the memory, or `false` if only one does.

`Drop` impl for `FixedSizeAllocator` checks the value of this field and only frees the memory if JS doesn't also hold a reference.

At present, `is_double_owned` is always `false`, because we're not transferring buffers over to JS yet. But it'll come into play in later PR. At that point the finalizer which runs when JS garbage collector collects the buffer will also update `is_double_owned`, and free the memory if Rust already dropped the `FixedSizeAllocator`, so JS is the only remaining owner.
